openmc/source.py

```diff
@@ -379,6 +393,139 @@ def from_xml_element(cls, elem: ET.Element, meshes=None) -> 'openmc.SourceBase':
 
         return source
 
+class MeshSource(SourceBase):
+    """A mesh-based source in which random positions are uniformly sampled
+    within mesh elements and each element can have independent angle, energy,
+    and time distributions. The element sampled is chosen based on the relative
+    strengths of the sources applied to the elements. The strength of the mesh
+    source as a whole is the sum of all source strengths applied to the elements.
+
+    Parameters
+    ----------
+    mesh : openmc.MeshBase
+        The mesh on which source sites will be generated.
+    sources : iterable of openmc.SourceBase
+        Sources for each element in the mesh. If spatial distributions are set
+        on any of the source objects, they will be ignored during source site
+        sampling.
+
+    Attributes
+    ----------
+    mesh : openmc.MeshBase
+        The mesh on which source sites will be generated.
+    sources : numpy.ndarray or iterable of openmc.SourceBase
+        The set of sources to apply to each element. The shape of this array must match
+        the shape of the mesh with and exception in the case of unstructured mesh, which
+        allows for application of 1-D array or iterable.
```

Comment:
Just a question on this part, if we have a mesh that has dimentions 2 by 2 by 2 then I understand we would pass in a 2 by 2 by 2 numpy array. If some of the mesh elements have no source because they are void space how would we do that, I guess we pass in a source with strength 0?

Reply:
I was thinking about this today as well and will do so more tomorrow. For now, yes you'd pass an `IndependentSource` object with a strength of zero.